inferno/callbacks.py

```diff
@@ -109,35 +111,70 @@ class AverageLoss(Callback):
       determine the correct average in case the batch sizes are not
       the same across all batches.)
 
+    keys_optional: list of str or None (default=None)
+      If not None, this should be a list of keys whose presence is
+      optional. By default, keys indicated in `key_sizes` are
+      mandatory, but sometimes we want optional keys. E.g., if keys
+      refer to validation data, but validation data is not always
+      present, those keys should be optional. When a key that is not
+      optional is missing, an exception will be raised.
+
     Attributes
     ----------
     default_key_sizes
       By default, average train loss and average valid loss are
       determined. If any of the losses is not present, it is
       ignored.
 
-    """
+      """
     default_key_sizes = {'train_loss': 'train_batch_size',
                          'valid_loss': 'valid_batch_size'}
 
-    def __init__(self, key_sizes=None):
+    def __init__(self, key_sizes=None, keys_optional=None):
         self.key_sizes = {} if key_sizes is None else key_sizes
 
+        if keys_optional is None:
+            self.keys_optional = []
+        elif isinstance(keys_optional, str):
+            self.keys_optional = [keys_optional]
+        else:
+            self.keys_optional = keys_optional
+
+        self._check_keys_duplicated()
+
+    def _is_optional(self, key):
+        return (key in self.keys_optional) or (key in self.default_key_sizes)
+
     def _yield_key_losses_bs(self, history):
-        for key_loss, key_size in self.default_key_sizes.items():
-            try:
-                row = history[-1, 'batches', :, (key_loss, key_size)]
-                yield key_loss, list(zip(*row))
-            except KeyError:
-                pass
-        for key_loss, key_size in self.key_sizes.items():
+        key_sizes = chain(
+            self.default_key_sizes.items(), self.key_sizes.items())
+        for key_loss, key_size in key_sizes:
+            if self._is_optional(key_loss) or self._is_optional(key_size):
+                try:
+                    row = history[-1, 'batches', :, (key_loss, key_size)]
+                    yield key_loss, list(zip(*row))
+                except KeyError:
+                    pass
+                continue
             row = history[-1, 'batches', :, (key_loss, key_size)]
             yield key_loss, list(zip(*row))
 
+    def _check_keys_duplicated(self):
+        duplicates = duplicate_items(self.default_key_sizes, self.key_sizes)
+        if duplicates:
+            raise ValueError("AverageLoss found duplicate keys: {}"
+                             "".format(', '.join(sorted(duplicates))))
+
+    def _check_keys_missing(self, history):
+        check_keys = []
+        for key in flatten(self.key_sizes.items()):
```

Comment:
Maybe add an explanation why we use `flatten(key_sizes.items())`, e.g.

    # Check value keys and size keys alike.